Add rt1180 lptmr support

lucien-nxp opened the pull request:
add 3 lptmr instances for RT118X series.
counter_basic_api case have been verified on my local for lptmr1 instance.

lucien-nxp committed:
dts: arm: nxp: rt118x: add lptmr instances

Config/Enable lptmr1/2/3 clock
Add 3 lptmr instances for RT118X

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
boards: nxp: mimxrt1180_evk: set lptmr1 status as OK

set lptmr1 status as OK
test counter_basic_api passed on cm33/cm7 cores

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented:
@hakehuang , please help verify counter_basic_api cases for lptmr1 instance.
Thank you.

danieldegrasse commented on boards/nxp/mimxrt1180_evk/doc/index.rst:
```diff
@@ -109,6 +109,8 @@ configuration supports the following hardware features:
 | ACMP      | on-chip    | analog comparator                   |
 +-----------+------------+-------------------------------------+
+| LPTMR     | on-chip    | counter                             |
++-----------+------------+-------------------------------------+
```
Need to add `-counter` to board YAML

danieldegrasse reviewed (commented):
Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR

hakehuang commented:
board testing pass

```
2024-09-09 09:49:52,310 - twister - DEBUG - DEVICE: *** Booting Zephyr OS build v3.7.0-2277-gd16f51f120e4 ***
2024-09-09 09:50:22,472 - twister - DEBUG - DEVICE: Running TESTSUITE counter_basic
2024-09-09 09:50:22,478 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:22,780 - twister - DEBUG - DEVICE: START - test_all_channels
2024-09-09 09:50:22,782 - twister - DEBUG - DEVICE: Testing gpt@2ec0000
2024-09-09 09:50:22,915 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:23,019 - twister - DEBUG - DEVICE: PASS - test_all_channels in 0.236 seconds
2024-09-09 09:50:23,025 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:23,029 - twister - DEBUG - DEVICE: START - test_cancelled_alarm_does_not_expire
2024-09-09 09:50:23,031 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:23,133 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:23,239 - twister - DEBUG - DEVICE: SKIP - test_cancelled_alarm_does_not_expire in 0.205 seconds
2024-09-09 09:50:23,245 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:23,247 - twister - DEBUG - DEVICE: START - test_late_alarm
2024-09-09 09:50:23,249 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:23,352 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:23,456 - twister - DEBUG - DEVICE: SKIP - test_late_alarm in 0.205 seconds
2024-09-09 09:50:23,462 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:23,464 - twister - DEBUG - DEVICE: START - test_late_alarm_error
2024-09-09 09:50:23,467 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:23,569 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:23,667 - twister - DEBUG - -- west flash: using runner linkserver
-- runners.linkserver: LinkServer: /usr/local/LinkServer/LinkServer, version v1.6.133

2024-09-09 09:50:23,673 - twister - DEBUG - DEVICE: SKIP - test_late_alarm_error in 0.206 seconds
2024-09-09 09:50:23,679 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:23,682 - twister - DEBUG - DEVICE: START - test_multiple_alarms
2024-09-09 09:50:23,684 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:23,787 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:23,891 - twister - DEBUG - DEVICE: SKIP - test_multiple_alarms in 0.205 seconds
2024-09-09 09:50:23,897 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:23,900 - twister - DEBUG - DEVICE: START - test_set_top_value_with_alarm
2024-09-09 09:50:23,904 - twister - DEBUG - DEVICE: E: Wrap can only be set to 0xffffffff
2024-09-09 09:50:23,906 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:24,008 - twister - DEBUG - DEVICE: Testing lptmr@4300000
2024-09-09 09:50:24,222 - twister - DEBUG - DEVICE: PASS - test_set_top_value_with_alarm in 0.318 seconds
2024-09-09 09:50:24,228 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:24,231 - twister - DEBUG - DEVICE: START - test_short_relative_alarm
2024-09-09 09:50:24,234 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:24,337 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:24,441 - twister - DEBUG - DEVICE: SKIP - test_short_relative_alarm in 0.206 seconds
2024-09-09 09:50:24,447 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:24,451 - twister - DEBUG - DEVICE: START - test_single_shot_alarm_notop
2024-09-09 09:50:24,453 - twister - DEBUG - DEVICE: Testing gpt@2ec0000
2024-09-09 09:50:24,625 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:24,730 - twister - DEBUG - DEVICE: PASS - test_single_shot_alarm_notop in 0.276 seconds
2024-09-09 09:50:24,736 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:24,739 - twister - DEBUG - DEVICE: START - test_single_shot_alarm_top
2024-09-09 09:50:24,743 - twister - DEBUG - DEVICE: E: Wrap can only be set to 0xffffffff
2024-09-09 09:50:24,745 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:24,848 - twister - DEBUG - DEVICE: Skipped for lptmr@4300000
2024-09-09 09:50:24,952 - twister - DEBUG - DEVICE: SKIP - test_single_shot_alarm_top in 0.209 seconds
2024-09-09 09:50:24,958 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:24,961 - twister - DEBUG - DEVICE: TESTSUITE counter_basic succeeded
2024-09-09 09:50:24,965 - twister - DEBUG - DEVICE: Running TESTSUITE counter_no_callback
2024-09-09 09:50:24,971 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:25,274 - twister - DEBUG - DEVICE: START - test_set_top_value_without_alarm
2024-09-09 09:50:25,278 - twister - DEBUG - DEVICE: E: Wrap can only be set to 0xffffffff
2024-09-09 09:50:25,280 - twister - DEBUG - DEVICE: Skipped for gpt@2ec0000
2024-09-09 09:50:25,383 - twister - DEBUG - DEVICE: Testing lptmr@4300000
2024-09-09 09:50:25,493 - twister - DEBUG - DEVICE: PASS - test_set_top_value_without_alarm in 0.215 seconds
2024-09-09 09:50:25,499 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:25,503 - twister - DEBUG - DEVICE: TESTSUITE counter_no_callback succeeded
2024-09-09 09:50:25,506 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY START ------
2024-09-09 09:50:25,515 - twister - DEBUG - DEVICE: SUITE PASS - 100.00% [counter_basic]: pass = 3, fail = 0, skip = 6, total = 9 duration = 2.066 seconds
2024-09-09 09:50:25,521 - twister - DEBUG - DEVICE: - PASS - [counter_basic.test_all_channels] duration = 0.236 seconds
2024-09-09 09:50:25,529 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 0.205 seconds
2024-09-09 09:50:25,535 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_late_alarm] duration = 0.205 seconds
2024-09-09 09:50:25,541 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_late_alarm_error] duration = 0.206 seconds
2024-09-09 09:50:25,548 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_multiple_alarms] duration = 0.205 seconds
2024-09-09 09:50:25,555 - twister - DEBUG - DEVICE: - PASS - [counter_basic.test_set_top_value_with_alarm] duration = 0.318 seconds
2024-09-09 09:50:25,561 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_short_relative_alarm] duration = 0.206 seconds
2024-09-09 09:50:25,568 - twister - DEBUG - DEVICE: - PASS - [counter_basic.test_single_shot_alarm_notop] duration = 0.276 seconds
2024-09-09 09:50:25,575 - twister - DEBUG - DEVICE: - SKIP - [counter_basic.test_single_shot_alarm_top] duration = 0.209 seconds
2024-09-09 09:50:25,585 - twister - DEBUG - DEVICE: SUITE PASS - 100.00% [counter_no_callback]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.215 seconds
2024-09-09 09:50:25,593 - twister - DEBUG - DEVICE: - PASS - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.215 seconds
2024-09-09 09:50:25,596 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY END ------
2024-09-09 09:50:25,602 - twister - DEBUG - DEVICE: ===================================================================
2024-09-09 09:50:25,606 - twister - DEBUG - DEVICE: RunID: b2b95436ba9de8dcf4aaf99693222a6f
2024-09-09 09:50:25,609 - twister - DEBUG - DEVICE: PROJECT EXECUTION SUCCESSFUL
```

lucien-nxp commented:
> Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR

I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development

decsny commented:
> > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> 
> I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development

soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

lucien-nxp commented:
> > > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> > 
> > 
> > I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development
> 
> soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

Updated code, and test blinky passed on my local when using lptmr as systick.
> > > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> > 
> > 
> > I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development
> 
> soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

Updated code, and test blinky passed on my local when using lptmr as systick.

> > > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> > 
> > 
> > I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development
> 
> soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

Updated code, and test blinky passed on my local when using lptmr as systick.

> > > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> > 
> > 
> > I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development
> 
> soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

lucien-nxp commented:
> > > Has this change been verified using the LPTMR as a kernel timer source? I see that we clock the LPTMR if `CONFIG_MCUX_LPTMR_TIMER` is set, but typically we also need to set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to a different value if using the LPTMR
> > 
> > 
> > I found that LPTMR can also be used as Timer, so I fill this macro in the soc.c where clock is enabled in advance, so you mean that I should delete it, right? What if I don't enable LPTMR as a timer? I just thought it would be easier to write this in advance for later development
> 
> soc/nxp/kinetis/ke1xz is an example of an soc which can use lptmr as system timer, and soc/nxp/imxrt/imxrt5xx is an example of an RT device which has two different system timer options enabled

Updated code.  Test blinky passed on my local when using lptmr as systick. However, still keep use 'nxp,lptmr' as lptmr's campatible in rt118x dtsi file.

danieldegrasse commented:
> Updated code. Test blinky passed on my local when using lptmr as systick. However, still keep use 'nxp,lptmr' as lptmr's campatible in rt118x dtsi file.

Just to be clear here- blinky was tested with an oscilloscope or similar watching the signal on the LED? I ask because we need to be sure the system is actually running at the clock frequency the Kconfig settings suggest, and one way to do this is to toggle an external signal (such as the LED in blinky) at a set time interval, and validate that the interval is correct using a second device with an independent time source, like an oscilloscope.

lucien-nxp commented:
> [ `d16f51f` ](/zephyrproject-rtos/zephyr/commit/d16f51f120e46f8ad0548dbf7b4a0db797fd895a)

OK, I will prepare another PR to use lptmr as systick, and validate that the interval with logic analyzer.

lucien-nxp commented:
> > Updated code. Test blinky passed on my local when using lptmr as systick. However, still keep use 'nxp,lptmr' as lptmr's campatible in rt118x dtsi file.
> 
> Just to be clear here- blinky was tested with an oscilloscope or similar watching the signal on the LED? I ask because we need to be sure the system is actually running at the clock frequency the Kconfig settings suggest, and one way to do this is to toggle an external signal (such as the LED in blinky) at a set time interval, and validate that the interval is correct using a second device with an independent time source, like an oscilloscope.

OK, I will prepare another PR to use lptmr as systick, and validate that the interval with logic analyzer.

danieldegrasse commented:
> OK, I will prepare another PR to use lptmr as systick, and validate that the interval with logic analyzer.

Has this been validated? It seems like this PR still enables using the LPTMR as a system clock source.
